Update to Orleans v1.5

wattsm opened the pull request:
This PR supersedes #13 and updates the test kit to work with Orleans v 1.5.

Orleans v1.5 makes `GrainCreator`, used by the test kit to instantiate grains, as `internal` meaning it cannot be used. I raised an issue on the Orleans repository about alternatives and was direct to [this thread](https://github.com/dotnet/orleans/issues/3086). The short answers seems to be that best way of unit testing a grain is to accept `IGrainRuntime` and `IGrainIdentity` in the constructor and let Orleans inject them for you in production. 

For the test kit this would mean either implementing some kind of primitive DI or using a container in conjunction with `TestServiceProvider`. The route of least resistance here seemed to be to create a `TestGrainCreator` type that mirrors the functionality of the internal `GrainCreator` using reflection. It is not pretty nor particularly robust, but the tests all pass. 

I'm happy to update the PR if anyone can suggest a better way of doing this?

sachokFoX commented:
Hi @wattsm !
There is one small issue with current version in this pull request.
Here is code from `TestGrainCreator.CreateGrainInstance()`:
```csharp
if(baseType.IsGenericType && baseType.GetGenericTypeDefinition() == typeof(Grain<>))
{
    //Set the state value
    ...
```

Issue is when your grain does not directly inherited from `Grain<T>` class. For example:
```csharp
public class BaseStateGrain<T> : Grain<T> {}

public class MyGrain<MyGrainState> : BaseStateGrain<MyGrainState> {}
```
Problem is here:
`baseType.GetGenericTypeDefinition() == typeof(Grain<>)`

So for example above `State` property will not be initialized and call to 
`TestKitSile.CreateGrain<T>() { ... _grainStateManager.AddState(grain, state); ... }` 
will fail with `null` state.

Maybe use recursion as in `TestKitSilo.GetProperty()` ?

wattsm commented:
Hi @sachokFoX - I thought that for a grain to be stateful in it needed to inherit from `Grain<T>`, or does it now work via convention as well, i.e. if you have a `State` property then Orleans will populate it?

sachokFoX commented:
Hi @wattsm,
> I thought that for a grain to be stateful in it needed to inherit from Grain<T>

you are right, but the problem I explained above is about when you have at least 2 layers of inheritance from `Grain<T>`.  `baseType.GetGenericTypeDefinition()` will return `BaseStateGrain<T>` and not `Grain<T>` in case of upper example.

wattsm commented:
Ah, yes - sorry, I misunderstood the point you were making. I'll tweak the PR shortly 👍

wattsm commented:
Hi @sachokFoX,

I've updated `TestGrainCreator` so that it now uses `TypeHelper.IsSubclassOfRawGeneric` which already existed. I've used `GetProperty` and moved it from `TestKitSilo` to `TypeHelper` so all the reflection is in the same place.

sachokFoX commented:
Hi @wattsm, thank you for update!

jamescarter-le commented:
If this PR is looking good to bring TestKit up to 1.5 could we get an ETA on the merge?

Thanks.

sachokFoX commented:
Hi @jamescarter-le, we can ask @dsarfati for help.
Also you can use artifact nuget from appveyor build if you have private nuget feed for your project.

jamescarter-le commented:
Thanks Bohdan I will use a private feed for now.

On Sun, 20 Aug 2017, 09:51 Bohdan Sachkovskyi <notifications@github.com>
wrote:

> Hi @jamescarter-le <https://github.com/jamescarter-le>, we can ask
> @dsarfati <https://github.com/dsarfati> for help.
> Also you can use artifact nuget from appveyor build if you have private
> nuget feed for your project.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/OrleansContrib/OrleansTestKit/pull/14#issuecomment-323572363>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABkUrLnaGiDDyOYsVLOkPX0DRKb6Au6Vks5sZ_N0gaJpZM4OQ8Gd>
> .
>

wattsm commented:
@sachokFoX @dsarfati What is the process for pushing this to nuget.org? I'm not familiar with AppVeyor - I thought it would publish the master branch automatically.

dsarfati commented:
I think it will automatically push to nuget.org, you will need to update the assembly version so that it appears as a new nuget version, otherwise it wont push. If you look at the .nuspec file you will see that the nuget version is using the assembly version. Go ahead and try to roll that 0.3.0 and let's see if it works automatically. If not, let me know and i can see if i can get the auto push working.

wattsm commented:
I updated the assembly version to `0.2.0` as a part of this PR, as I assumed that'd be how it worked. I notice the AppVeyor config has a version mask of `0.1.{build}` - does this need to change to `0.2.{build}`?

sachokFoX commented:
Hi @wattsm, I am also not familiar with appveyor :( I found this documentation https://www.appveyor.com/docs/deployment/nuget/ and as I understand you need to add `deploy:` section to `appveyor.yml` with your `api_key` but is secure info... Maybe you need access to appveyor project and update nuget info.
